core/drivers/crypto/caam/caam_jr.c

```diff
@@ -0,0 +1,781 @@
+		 * also touching it
+		 */
+		cpu_spin_lock(&jr_privdata->callers_lock);
+		for ((idx_jr = 0), (found = (-1));
+				(idx_jr < jr_privdata->nbJobs) && (found < 0);
+				idx_jr++) {
+			/*
+			 * Search for the caller information corresponding to
+			 * the completed JR.
+			 * Don't use the outread_index or inwrite_index because
+			 * completion can be out of order compared to input
+			 * buffer
+			 */
+			caller = &jr_privdata->callers[idx_jr];
+			if (caam_read_val(&jr_out->desc) == caller->pdesc) {
+				jobctx         = caller->jobctx;
+				jobctx->status = caam_read_val(&jr_out->status);
+
+				/* Update return Job Id mask */
+				if (caller->jobid & waitJobIds)
+					retJobId |= caller->jobid;
+
+				JR_TRACE("JR id=%d, context @0x%08"PRIxVA"",
+					caller->jobid, (vaddr_t)jobctx);
+				/* Clear the Entry Descriptor DMA */
+				caller->pdesc = 0;
+				caller->jobid = JR_JOB_FREE;
+				found = idx_jr;
+				JR_TRACE("Free space #%d in the callers array",
+					idx_jr);
+			}
+		}
+		cpu_spin_unlock(&jr_privdata->callers_lock);
+
+		/*
+		 * Remove the JR from the output list even if no
+		 * JR caller found
+		 */
+		hal_jr_del_job(jr_privdata->baseaddr);
+
+		/*
+		 * Increment index to next JR output entry taking care that
+		 * is a circular buffer of nbJobs size.
+		 */
+		jr_privdata->outread_index++;
+		jr_privdata->outread_index %= jr_privdata->nbJobs;
+
+		if ((found >= 0) && (jobctx->callbk)) {
+			/* Finally, execute user's callback */
+			jobctx->callbk(jobctx);
+		}
+
+	} while (--nbJobs_done);
+
+	cpu_spin_unlock_xrestore(&jr_privdata->inlock, exceptions);
+
+	return retJobId;
+}
+
+/*
+ * Enqueues a new job in the Job Ring input queue. Keep the caller's
+ * job context in private array.
+ *
+ * Inputs:
+ * jobctx   Caller's job context
+ *
+ * Outputs:
+ * jobId    Job Id enqueued
+ *
+ * Returns:
+ * CAAM_NO_ERROR  Success
+ * CAAM_BUSY      CAAM is busy
+ */
+static enum CAAM_Status do_jr_enqueue(struct jr_jobctx *jobctx, uint32_t *jobId)
+{
+	enum CAAM_Status retstatus = CAAM_BUSY;
+
+	struct caller_info *caller = NULL;
+	uint32_t exceptions = 0;
+	uint32_t job_mask = 0;
+	uint8_t  idx_jr = 0;
+	bool     found = false;
+
+	exceptions = cpu_spin_lock_xsave(&jr_privdata->inlock);
+
+	/* Stay lock until a job is available */
+	/* Check if there is an available JR index in the HW */
+	while (hal_jr_read_nbSlotAvailable(jr_privdata->baseaddr) == 0) {
+		/*
+		 * WFE will be exit be a SEV generated but the
+		 * interrupt handler or by a spin_unlock
+		 */
+		wfe();
+	};
+
+	/*
+	 * There is a space free in the input ring but it doesn't mean
+	 * that the job pushed is completed.
+	 * Completion is out of order. Look for a free space in the
+	 * caller data to push them and get a job id for the completion
+	 *
+	 * Lock the caller information array because dequeue is
+	 * also touching it
+	 */
+	cpu_spin_lock(&jr_privdata->callers_lock);
+	for ((idx_jr = 0), (found = false);
```

Comment:
With this loop in a helper function returning the found `struct caller_info *` would simplify things a bit.